src/utils/XFDR.cpp

```diff
@@ -0,0 +1,1095 @@
+
+    // Score range for calculating the FPs
+    static const double fpnum_score_step;
+
+    TOPPXFDR() :
+      TOPPBase("XFDR", "Calculates false discovery rate estimates on cross-link identifications", false)
+    {
+    }
+
+  protected:
+
+    /**
+     * @brief Used to define how PeptideIdentification are to be sorted based on some meta value (usually some score)
+     */
+    struct less_than_by_key
+    {
+      //Indizes of the rank one elements within the all_ids vector
+      const vector < PeptideIdentification > & all_ids;
+      const vector< UInt >  & rank_one_ids;
+      const String & meta_value;  // Meta value to sort the peptideIdentifications
+
+      inline bool operator()(const UInt & index1, const UInt & index2)
+      {
+        return (  TOPPXFDR::getXLMetaValue<double>(meta_value, all_ids[rank_one_ids[index1]], true)
+                > TOPPXFDR::getXLMetaValue<double>(meta_value, all_ids[rank_one_ids[index2]], true));
+      }
+    };
+
+    /**
+     * @brief Tests for descending ordering of a vector of rank one ids based on the score. Only used for assertions.
+     * @param order vector of ranks
+     * @param all_ids Vector with the actual PeptideIdentifictions which are queried for the score.
+     * @param rank_one_ids Vector of indices pointing to the rank one IDs in all_ids.
+     * @return Whether rank vector order reflects descending ordering of rank_one_ids by score.
+     */
+    static bool isSortedDescending(vector< UInt > & order, vector< PeptideIdentification > & all_ids, vector< UInt > & rank_one_ids )
+    {
+      for (Size i = 0; i < order.size() - 1; ++i)
+      {
+        if (  getXLMetaValue<double>("OpenXQuest:score", all_ids[rank_one_ids[order[i]]], true)
+            < getXLMetaValue<double>("OpenXQuest:score", all_ids[rank_one_ids[order[i + 1]]], true))
+        {
+          return false;
+        }
+      }
+      return true;
+    }
+
+    /**
+     * Searches for a meta value in the cross-link peptide ID. First in PeptideIdentification itself, then
+     * in the associated Peptide hits.
+     */
+    template<typename T>
+    static T getXLMetaValue(const String & name, const PeptideIdentification & pep_id, bool is_score)
+    {
+      if( pep_id.metaValueExists(name))
+      {
+        return static_cast<T>(pep_id.getMetaValue(name));
+      }
+      const vector< PeptideHit > & pep_hits = pep_id.getHits();
+      if (is_score)
+      {
+        return static_cast<T>(pep_hits[0].getScore());
+      }
+      if (pep_hits[0].metaValueExists(name))
+      {
+        return static_cast<T>(pep_hits[0].getMetaValue(name));
+      }
+      if (pep_hits.size() == 2 && pep_hits[1].metaValueExists(name))
+      {
+        return static_cast<T>(pep_hits[1].getMetaValue(name));
+      }
+      return DataValue::EMPTY;
+    }
+
+    /**
+     * @brief Prepares read vector of PeptideIdentification such that it can be further processed downstream
```

Comment:
maybe a bit more information what happens inside. the comment does not contain much additional information compared to the function name ;)